tests/src/gui/testqgsdoublevalidator.cpp

```diff
@@ -0,0 +1,138 @@
+/***************************************************************************
+    testqgsdoublespinbox.cpp
+     --------------------------------------
+    Date                 : June 2020
+    Copyright            : (C) 2020 Sebastien Peillet
+    Email                : sebastien dot peillet at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgstest.h"
+
+#include "qgsdoublevalidator.h"
+#include <QLineEdit>
+
+class TestQgsDoubleValidator: public QObject
+{
+    Q_OBJECT
+  private slots:
+    void initTestCase(); // will be called before the first testfunction is executed.
+    void cleanupTestCase(); // will be called after the last testfunction was executed.
+    void init(); // will be called before each testfunction is executed.
+    void cleanup(); // will be called after every testfunction.
+
+    void validate();
+    void toDouble();
+
+  private:
+
+};
+
+void TestQgsDoubleValidator::initTestCase()
+{
+
+}
+
+void TestQgsDoubleValidator::cleanupTestCase()
+{
+}
+
+void TestQgsDoubleValidator::init()
+{
+}
+
+void TestQgsDoubleValidator::cleanup()
+{
+}
+
+void TestQgsDoubleValidator::validate()
+{
+    QLocale().setDefault(QLocale::English);
```

Comment:
Can you please add test cases with group (thousands) separators?

I think we should test and accept all valid inputs inputs the user locale but also accept the C locale valid inputs as a fallback, for example assuming locale is it_IT, these would be all valid:

1,234  ( interpreted as 1.234 )
1.234 (interpreted as 1234 ... this is ambiguous but the user locale takes precedence)
1'234,0000001 ( 1234.0000001 )
1234,0000001 ( 1234.0000001 )
the following are interpreted using the fallback C locale:
1234.0000001  ( 1234.0000001  because it's valid in the fallback C locale)
1,234.0000001 ( 1234.0000001 because it's valid in the fallback C locale and it's not ambiguous)
1,234.000  ( 1234.0000001 because it's valid in the fallback C locale and it's not ambiguous)

Reply:
I add tests with group separator, as you said there is some ambiguous cases, in particular when C locale and user locale share the same decimalpoint and/or group separator. To handle it, the test [changes the expected value](https://github.com/SebastienPeillet/QGIS/commit/a52df08fc944e0563c5525fc2ba9b2422a349909#diff-cb725c98171908938554db1067e951f9R136) (validate method will return QValidator::Intermediate and not QValidator::Invalid).

Tell me if you think there is still missing tests